Document UDP component (#3918)
OttoWinter/esphomedocs: components/udp.rst

        - id: sensor_id2
          broadcast_id: different_id

      providers:
        - name: some-device-name
          encryption: "REPLACEME with some key"

    sensor:
      - platform: udp
        provider: some-device-name
        id: local_sensor_id
        remote_id: some_sensor_id

    binary_sensor:
      - platform: udp
        provider: unencrypted-device
        id: other_binary_sensor_id  # also used as remote_id

Configuration variables:
------------------------

- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
- **update_interval** (*Optional*, :ref:`config-time`): Interval between full broadcasts. Defaults to 15s.
- **port** (*Optional*, int): The destination UDP port number to use. Defaults to ``18511``.
- **addresses** (*Optional*, list of IPv4 addresses): One or more IP addresses to broadcast data to. Defaults to ``255.255.255.255``
  which is the local network broadcast address.
- **sensors** (*Optional*, list): A list of sensor IDs to be broadcast. Each entry may be just the sensor id, or may set a different id to be broadcast.

  - **id** (**Required**, :ref:`config-id`): The id of the sensor to be used

>   - **broadcast_id** (*Optional*, :ref:`config-id`): The id to be used for this sensor in the broadcast. Defaults to the same as the internal id.

Reviewer: ```suggestion
  - **broadcast_id** (*Optional*, string): The id to be used for this sensor in the broadcast. Defaults to the same as the internal id.
```

> - **binary_sensors** (*Optional*, list): A list of binary sensor IDs to be broadcast.

Reviewer: ```suggestion
- **binary_sensors** (*Optional*, list): A list of binary sensor IDs to be broadcast.

  - **id** (**Required**, :ref:`config-id`): The id of the binary sensor to be used
  - **broadcast_id** (*Optional*, string): The id to be used for this binary sensor in the broadcast. Defaults to the same as the internal id.

```

- **ping_pong_recycle_time** (*Optional*, :ref:`config-time`): Controls how often the ping-pong key is regenerated. Requires ``ping_pong_enable`` to be set. Defaults to 10 minutes. Can be set only on the consumer side.
- **providers** (*Optional*, list): A list of provider device names and optionally their secret encryption keys.

  - **name** (**Required**, string): The device name of the provider.
  - **encryption** (*Optional*, string): The provider's encryption key.

Wherever a provider name is required, this should be the node name configured in the ``esphome:`` block.

This component supports multiple configurations, making it possible to differentiate between consumers when providing data to them.
When receiving data in such a configuration, sensors need an ``udp_id`` configuration item to know where to expect data to come from.

Reliability
-----------

UDP, like any other network protocol, does not provide a guarantee that data will be delivered, but unlike TCP it does not
even provide any indication whether data has been successfully delivered or not. When any of the configured sensors changes state,

> the component will broadcast that sensor's state, but since this may not be d by a consumer, the UDP component

Reviewer: Not sure what the `d` is supposed to be here?
```suggestion
the component will broadcast that sensor's state, but since this may not be delivered to a consumer, the UDP component
```

also broadcasts *all* sensor data on a timed schedule, set by ``update_interval``. Even this does not guarantee

> delivery, but in practice unless the network has failed, updates will eventually be d, albeit possibly after

Reviewer: Another `d`? Delivered?
```suggestion
delivery, but in practice unless the network has failed, updates will eventually be delivered, albeit possibly after
```
